src/components/PVGenerator/PVGeneratorComponent.tsx

```diff
@@ -0,0 +1,233 @@
+            const endPoint = this.widgetStore.effectiveRegion.controlPoints[1];
+            const width = this.widgetStore.effectiveFrame.frameInfo.fileInfoExtended.width;
+            const height = this.widgetStore.effectiveFrame.frameInfo.fileInfoExtended.height;
+            const imageCorners: Point2D[] = [
+                {x: 0, y: 0},
+                {x: width - 1, y: 0},
+                {x: 0, y: height - 1},
+                {x: width - 1, y: height - 1}
+            ];
+
+            // check if image corners are on the same side of the line region. from https://stackoverflow.com/a/1560510
+            let sideValue = 0;
+            for (let corner of imageCorners) {
+                sideValue = sideValue + Math.sign((endPoint.x - startPoint.x) * (corner.y - startPoint.y) - (endPoint.y - startPoint.y) * (corner.x - startPoint.x));
+            }
+
+            if (sideValue <= 2 && sideValue >= -2) {
+                // check if start/end points are outside image
+                if ((startPoint.x < 0 && endPoint.x < 0) || (startPoint.y < 0 && endPoint.y < 0) || (startPoint.x > width && endPoint.x > width) || (startPoint.y > height && endPoint.y > height)) {
+                    return false;
+                }
+
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @computed get isLineInOnePixel(): boolean {
+        if (this.widgetStore.effectiveRegion) {
+            const startPoint = this.widgetStore.effectiveRegion.controlPoints[0];
+            const endPoint = this.widgetStore.effectiveRegion.controlPoints[1];
+            if (Math.round(startPoint.x) === Math.round(endPoint.x) && Math.round(startPoint.y) === Math.round(endPoint.y)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    constructor(props: WidgetProps) {
+        super(props);
+        makeObservable(this);
+
+        const appStore = AppStore.Instance;
+        // Check if this widget hasn't been assigned an ID yet
+        if (!props.docked && props.id === PVGeneratorComponent.WIDGET_CONFIG.type) {
+            // Assign the next unique ID
+            const id = appStore.widgetsStore.addPVGeneratorWidget();
+            appStore.widgetsStore.changeWidgetId(props.id, id);
+        } else {
+            if (!appStore.widgetsStore.pvGeneratorWidgets.has(this.props.id)) {
+                console.log(`can't find store for widget with id=${this.props.id}`);
+                appStore.widgetsStore.pvGeneratorWidgets.set(this.props.id, new PVGeneratorWidgetStore());
+            }
+        }
+    }
+
+    @action private onResize = (width: number, height: number) => {
+        this.width = width;
+        this.height = height;
+    };
+
+    private handleFrameChanged = (changeEvent: React.ChangeEvent<HTMLSelectElement>) => {
+        const appStore = AppStore.Instance;
+        if (appStore.activeFrame) {
```

Comment:
shall we check for effectiveFrame instead of activeFrame?

Reply:
done.